### PR TITLE
feat: [CI-4865]: Inject environment variables in Matrix steps

### PR DESCRIPTION
**Pull request by hemanthmantri**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/35915)

### CODE REVIEW

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/IntegrationStageUtils.java [hunk 1]
```diff
@@ -418,6 +428,59 @@ public static List<StepElementConfig> getAllSteps(List<ExecutionWrapperConfig> e
...
+      return;
+    }
+    // Read the envVariables from config, and inject the following in envVars:
+    // HARNESS_NODE_INDEX=<+strategy.iteration>
+    // HARNESS_NODE_TOTAL=<+strategy.iterations>
+    // config -> step -> _children -> spec -> has YAML in JSON string format.
+    if (config.getStep() != null && !config.getStep().isNull()) {
+      StepElementConfig stepConfig = getStepElementConfig(config);
+      // If there's no matrix strategy, nothing to do.
+      StrategyConfig strategyConfig = stepConfig.getStrategy();
```
Comment (rutvijmehta-harness) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/IntegrationStageUtils.java:442]: Do you need to have this condition or there are cases when they call be null?

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/IntegrationStageUtils.java [hunk 2]
```diff
@@ -418,6 +428,59 @@ public static List<StepElementConfig> getAllSteps(List<ExecutionWrapperConfig> e
     return stepElementConfigs;
   }
 
+  public static void injectStepEnvVariables(ExecutionWrapperConfig config) {
```
Comment (rutvijmehta-harness) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/IntegrationStageUtils.java:431]: Make the method name bit more specific about injecting variables for parallelism

**Review by rutvijmehta-harness — COMMENTED** [review]
LGTM apart from comments clean up and UTs

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Failed_** 
 


            
 #### 2 Issues 
* Code issue 2 Code issues 

 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/35915)

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/IntegrationStageUtils.java (hunk 1 above)
Comment (hemanthmantri) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/IntegrationStageUtils.java]: Removed this. With no strategy the env variables will be null but I've asked Sahil to inject (0, 1) if no parallelism is set.

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/IntegrationStageUtils.java (hunk 2 above)
Comment (hemanthmantri) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/IntegrationStageUtils.java]: Done

**Comment by hemanthmantri** [comment]
trigger messagecheck

**Review by rutvijmehta-harness — APPROVED** [review]
> Reviewed 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35915)* status: :shipit: complete! ✓

**Comment by hemanthmantri** [comment]
trigger messagecheck

**Review by rutvijmehta-harness — APPROVED** [review]
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35915)* status: :shipit: complete! ✓

**Pull request merged**